src/ripple/app/ledger/impl/TimeoutCounter.h

```diff
@@ -0,0 +1,142 @@
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_LEDGER_TIMEOUTCOUNTER_H_INCLUDED
+#define RIPPLE_APP_LEDGER_TIMEOUTCOUNTER_H_INCLUDED
+
+#include <ripple/app/main/Application.h>
+#include <ripple/beast/clock/abstract_clock.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/core/Job.h>
+#include <boost/asio/basic_waitable_timer.hpp>
+#include <mutex>
+
+namespace ripple {
+
+/**
+    This class is an "active" object. It maintains its own timer
+    and dispatches work to a job queue. Implementations derive
+    from this class and override the abstract hook functions in
+    the base.
+
+    This class implements an asynchronous loop:
+
+    1. The entry point is `setTimer`.
+
+    2. After `mTimerInterval`, `queueJob` is called, which schedules a job to
+       call `invokeOnTimer` (or loops back to setTimer if there are too many
+       concurrent jobs).
+
+    3. The job queue calls `invokeOnTimer` which either breaks the loop if
+       `isDone` or calls `onTimer`.
+
+    4. `onTimer` is the only "real" virtual method in this class. It is the
+       callback for when the timeout expires. Generally, its only responsibility
+       is to set `mFailed = true`. However, if it wants to implement a policy of
+       retries, then it has a chance to just increment a count of expired
+       timeouts.
+
+    5. Once `onTimer` returns, if the object is still not `isDone`, then
+       `invokeOnTimer` sets another timeout by looping back to setTimer.
+
+   This loop executes concurrently with another asynchronous sequence,
+   implemented by the subtype, that is trying to make progress and eventually
+   set `mComplete = true`. While it is making progress but not complete, it
+   should set `mProgress = true`, which is passed to onTimer so it can decide
+   whether to postpone failure and reset the timeout. However, if it can
+   complete all its work in one synchronous step (while it holds the lock), then
+   it can ignore `mProgress`.
+
+   This class is split from the PeerSet class.  // TODO remove after PR
+*/
+class TimeoutCounter
+{
+protected:
+    using ScopedLockType = std::unique_lock<std::recursive_mutex>;
```

Comment:
can we make this a non-recursive mutex?

Reply:
The new derived classes don't need `recursive_mutex`. I looked at the existing ones quickly, i.e. `InboundLedger` and `TransactionAcquire`. I didn't see they needed it. But I don't know enough of those classes and how to test those to make the change.